client/metalink.c

```diff
@@ -0,0 +1,394 @@
+            BAIL_ON_TDNF_ERROR(dwError);
+        } 
+        else 
+        {
+            ml_url_info->preference = prefValue;
+        }
+        value = NULL;
+    }
+
+    //Get URL Content.
+    xmlContValue = xmlNodeGetContent(node);
+    if(xmlContValue == NULL)
+    {
+        dwError = ERROR_TDNF_ML_PARSER_MISSING_URL_CONTENT;
+        pr_err("URL is no present in URL element", value);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    value = (const char*)xmlContValue;
+    dwError = TDNFAllocateString(value, &(ml_url_info->url));
+    BAIL_ON_TDNF_ERROR(dwError);
+
+cleanup:
+    if(xmlContValue != NULL)
+    {
+        xmlFree(xmlContValue);
+        xmlContValue = NULL;  
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFXmlParseData(
+    TDNF_ML_CTX *ml_ctx,
+    xmlNode *node,
+    const char *filename
+    )
+{
+    uint32_t dwError = 0;
+    xmlChar* xmlContValue = NULL;
+    char *size = NULL;
+    char *nptr = NULL;
+
+    if(!ml_ctx || !node || IsNullOrEmptyString(filename))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    //Looping through all the nodes from root and parse all children nodes.
+    while(node)
+    {
+        if(node->type == XML_ELEMENT_NODE)
+        {                                                       
+            if(!strcmp((const char*)node->name, TAG_NAME_FILE))
+            {                     
+                dwError = TDNFParseFileTag(ml_ctx, node, filename);
+                BAIL_ON_TDNF_ERROR(dwError);
+            }                     
+            else if(!strcmp((const char*)node->name, TAG_NAME_SIZE))
+            {                     
+                //Get File Size.
+                xmlContValue = xmlNodeGetContent(node);
+                if(xmlContValue == NULL)
+                {
+                    dwError = ERROR_TDNF_ML_PARSER_MISSING_FILE_SIZE;
+                    pr_err("XML Parser Error:File size is missing: %s", size);
+                    BAIL_ON_TDNF_ERROR(dwError);
+                }
+                size = (char*)xmlContValue;
+                ml_ctx->size = strtoll(size, &nptr, 0);
+                xmlFree(xmlContValue);
+            }                     
+            else if(!strcmp((const char*)node->name, TAG_NAME_HASH))
+            {                     
+                dwError = TDNFParseHashTag(ml_ctx, node);
+                BAIL_ON_TDNF_ERROR(dwError);
+            }                     
+            else if(!strcmp((const char*)node->name, TAG_NAME_URL))
+            {                                                     
+                dwError = TDNFParseUrlTag(ml_ctx, node);
+                BAIL_ON_TDNF_ERROR(dwError);
+            }
+        }                       
+        TDNFXmlParseData(ml_ctx, node->children, filename);
+        node = node->next;
+    }
+    BAIL_ON_TDNF_ERROR(dwError);
```

Comment:
pls explain why we are checking error here?

Reply:
Removed.